### PR TITLE
Add validation webhook for service profiles

### LINKED ISSUE #2075: Improve validation of service profiles
The built-in [validation available to CRDs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation) is very limited.  In particular, it doesn't allow us to validate:

* that there are no unrecognized fields
* the recursive structure of `condition`s

This means that is it a common pitfall to accidentally use an incorrect field name in a condition.  If this happens, routes will not match correctly but it will be unclear why.

We can do more thorough validation by adding a validating admission webhook for service profiles.

### PR DESCRIPTION
**Pull request by alpeb**
Fixes #2075

The new webhook endpoint is created in a new container in the
linkerd-controller pod.

The proxy injector webhook was refactored to extract-out the scaffolding
to be reused by the validation webhook (and eventually by any other
admission webhooks we'd like to create in the future). The new common
stuff now reside under `pkg/webhook`.

The validation webhook simply sets up the endpoint server and the
webhook config using `pkg/webhook`, and calls the old
`profiles.Validate` to perform the actual validation.

I Also got rid of the unused `-no-init-container` and `-tlsEnabled`
flags in the proxy injector.

Todo in this PR: tests.

Todo in a follow-up PRs: remove the SP check from the CLI check.

### CODE REVIEW

**Comment by grampelberg** [comment]
@alpeb cool! Looks like it does a pretty good job explaining what broke as well =)

**Comment by alpeb** [comment]
@adleong the dependencies changes were caused by having mistakingly imported Prometheus's logger instead of the usual one. I fixed that along with tests.

File: controller/sp-validator/webhook.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// AdmitSP verifies that the received Admission Request contains a valid
+// Service Profile definition
+func AdmitSP(
+	_ kubernetes.Interface,
+	request *admissionv1beta1.AdmissionRequest,
+) (*admissionv1beta1.AdmissionResponse, error) {
+	admissionResponse := &admissionv1beta1.AdmissionResponse{Allowed: true}
+	if err := profiles.Validate(request.Object.Raw); err != nil {
+		admissionResponse.Allowed = false
+		admissionResponse.Result = &metav1.Status{Message: err.Error()}
```
Comment (siggy) [inline controller/sp-validator/webhook.go]: nice!

File: pkg/webhook/server.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
+package webhook
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/webhook/server.go]: You're right, I'll move all that over `controller/webhook`.

File: controller/cmd/sp-validator/main.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
+package main
```
[the remark this answers is not in the document]
Comment (alpeb) [reply controller/cmd/sp-validator/main.go]: I've refactored both `main.go` s so they just set up the config and use a new common `launcher.go`.
Both `webhook_ops.go` have a similar structure but use different types all over, so I think it's better to leave them both. Also it seems to me both template files are different enough to leave them as separate files.
As for the tests, `webhook_test.go` verifies that the auto injection works as expected.
The service profile validator is really just a wrapper around `profiles.Validate()` which already has unit tests inside its package. The wrapping part is also already tested a bit in `server_test.go``

[the remark this answers is not in the document]
Comment (olix0r) [reply controller/cmd/sp-validator/main.go]: It's worth noting that much of the bootstrapping stuff is intended to be pretty-short lived. In 2.4, we plan to move webhook creation into the installation process.

File: controller/proxy-injector/webhook.go [hunk 4]
```diff
@@ -97,7 +33,7 @@ func (w *Webhook) inject(request *admissionv1beta1.AdmissionRequest) (*admission
 		return nil, err
 	}
 
-	namespace, err := w.client.CoreV1().Namespaces().Get(request.Namespace, metav1.GetOptions{})
+	namespace, err := client.CoreV1().Namespaces().Get(request.Namespace, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (alpeb) [reply controller/proxy-injector/webhook.go]: Good point :+1:  I've issued #2566 to track that work.

**Comment by alpeb** [comment]
I've addressed all the feedback in my latest push. But thinking on rebasing after #2555, #2547 and #2500 land because this is gonna be painful, so no need to endure that pain in those branches as well :hurtrealbad:

**Comment by alpeb** [comment]
Superseded by #2623

**Pull request closed without merging**